sync/fetch_queue.go

```diff
@@ -0,0 +1,315 @@
+}
+
+type fetchJob struct {
+	items []Item
+	ids   []types.Hash32
+}
+
+//todo make the queue generic
+type fetchQueue struct {
+	log.Log
+	BatchRequestFactory
+	*sync.Mutex
+	workerInfra WorkerInfra
+	pending     map[types.Hash32][]chan bool
+	handle      func(fj fetchJob)
+	checkLocal  CheckLocalFunc
+	queue       chan []types.Hash32 //types.TransactionId //todo make buffered
+}
+
+func (vq *fetchQueue) Close() {
+	vq.Info("done")
+	close(vq.queue)
+}
+
+//todo batches
+func (fq *fetchQueue) work() error {
+	output := fetchWithFactory(NewFetchWorker(fq.workerInfra, 1, fq.BatchRequestFactory, fq.queue))
+	for out := range output {
+		if out == nil {
+			fq.Info("close queue")
+			return nil
+		}
+
+		bjb, ok := out.(fetchJob)
+		if !ok {
+			fq.Error(fmt.Sprintf("Type conversion err %v", out))
+			continue
+		}
+
+		if len(bjb.ids) == 0 {
+			return fmt.Errorf("channel closed")
+		}
+
+		if len(bjb.items) == 0 {
+			fq.Warning("could not fetch any items ") //todo print ids
+			continue
+		}
+
+		fq.handle(out.(fetchJob))
+		fq.Debug("next batch")
+	}
+	return nil
+}
+
+func (fq *fetchQueue) addToPendingGetCh(ids []types.Hash32) chan bool {
+	return getDoneChan(fq.addToPending(ids))
+}
+
+func (fq *fetchQueue) addToPending(ids []types.Hash32) []chan bool {
+	fq.Lock()
+	deps := make([]chan bool, 0, len(ids))
+	var idsToAdd []types.Hash32
+	for _, id := range ids {
+		ch := make(chan bool, 1)
+		deps = append(deps, ch)
+		if _, ok := fq.pending[id]; !ok {
+			idsToAdd = append(idsToAdd, id)
+		}
+		fq.pending[id] = append(fq.pending[id], ch)
+	}
+	fq.Unlock()
+	fq.queue <- idsToAdd
+	return deps
+}
+
+func (fq *fetchQueue) invalidate(id types.Hash32, valid bool) {
+	fq.Debug("done with %v !!!!!!!!!!!!!!!! %v", id.ShortString(), valid)
+	fq.Lock()
+	deps := fq.pending[id]
+	delete(fq.pending, id)
+	fq.Unlock()
+
+	for _, dep := range deps {
+		dep <- valid
+	}
+}
+
+//returns txs out of txids that are not in the local database
+func (tq *fetchQueue) Handle(itemIds []types.Hash32) ([]Item, error) {
+	if len(itemIds) == 0 {
+		tq.Debug("handle empty item ids slice")
+		return nil, nil
+	}
+
+	unprocessedItems, processedItems, missing := tq.checkLocal(itemIds)
+	if len(missing) > 0 {
+
+		output := tq.addToPendingGetCh(missing)
+
+		if success := <-output; !success {
+			return nil, errors.New(fmt.Sprintf("could not fetch all items"))
+		}
+
+		unprocessedItems, processedItems, missing = tq.checkLocal(itemIds)
+		if len(missing) > 0 {
+			return nil, errors.New("could not find all items even though fetch was successful")
+		}
+	}
+
+	txs := make([]Item, 0, len(itemIds))
+	for _, id := range itemIds {
+		if tx, ok := unprocessedItems[id]; ok {
+			txs = append(txs, tx)
+		} else if _, ok := processedItems[id]; ok {
+			continue
+		} else {
+			return nil, errors.New(fmt.Sprintf("item %v was not found after fetch was done", hex.EncodeToString(id[:])))
+		}
+	}
+	return txs, nil
+}
+
+//todo make the queue generic
+type txQueue struct {
```

Comment:
It is possible that you can avoid txQueue (and atxQueue).
1. fetchQueue can recv items that implements the func Hash32 (then, you don't need the cast in HandleTxs for example)
2. The fetchQueue.handle function can be constructed, with its dependencies, prior to the construction of the fetchQueue and passed to it. Then, the handle function signature can be simple and type-free as its dependencies have already been injected to it.

Look into it to see if it is possible as it can unburden the need in extra structs.

Reply:
because the invalidate method is dependent on the queue instance it creates more mess than it saves right now
but after we unify the ids this will all disappear

Reply:
Ok cool. Either way, it is cleaner than before.